### ISSUE TITLE
Implement new Rest API function introduced in WP5.0.0

### ISSUE THREAD

**Issue by zulfgani** [ClassicPress/ClassicPress#401]
Function name: `rest_preload_api_request` located in rest-api.php
## Expected behavior
Installed the new Assistant plugin and on activation it should just run on the frontend.

## Current behavior
However, it throws a WARNING that the function `rest_preload_api_request` is not found or invalid function name

## Possible solution
`rest_preload_api_request` was introduced in WP v5.0.0 and therefore needs to ported over to CP.

File: wp-includes/rest-api.php ~ Line: 1348

## Steps to reproduce (for bugs)
1. Install Assistant and activate
2. Check frontend to see the WARNING

## Context
Just evaluating the new plugin for suitability with ClassicPress before recommending it ~ the missing function may put a dumper on it.

**Comment by Mte90** [comment]
It is probably on https://github.com/ClassicPress/ClassicPress/issues/350
The big question that I was talking already with @nylen is how we want to keep the compatibility with wp 4.9?
Because add new functions doesn't mean backward compatibility as code but only as bugfix as example?
This enters a big discussion about how CP want to keep the compatibility in the various areas like bugfixing, code consistency and features.

**Comment by nylen** [comment]
>Installed the new Assistant plugin

@zulfgani : what plugin is this?  Can we get a link or some more information about what it does?

What does the `rest_api_preload_request` function do?  Most helpful is to leave a link to the code where everyone can see it easily.

This particular error is also a problem with this plugin in WP 4.9.x, so you could also report it as such to the plugin author.

In general I have no problem with merging functions like this over to ClassicPress, but we do need to be careful.  We advertise ourselves as `$wp_version = '4.9.10'`, and if we deviate from this it needs to be worth it.

Also, regardless of where it comes from, code that is new to ClassicPress needs to be tested, explained and understood before it can be merged into core.

**Comment by nylen** [comment]
>What does the `rest_api_preload_request` function do? Most helpful is to leave a link to the code where everyone can see it easily.

Like this:  https://github.com/WordPress/wordpress-develop/blob/5.1.1/src/wp-includes/rest-api.php#L1338-L1404

From there, it is a good idea to look at the "blame" view that shows the history of this code:

![2019-03-16T16 14 43-05 00](https://user-images.githubusercontent.com/227022/54481740-a95da100-4806-11e9-988c-ca3307df3a11.png)

This points towards several different commits and several different Trac tickets.  As a next step for collecting more information, it would be helpful to add links to the related tickets and changesets here on this issue, with a brief description of what each one is doing.

**Comment by nylen** [comment]
I have been thinking more about how we decide whether to accept changes from WP. We can't accept all of them, there are just too many and a lot of them wouldn't make sense anyway. I did come up with some more concrete criteria though:

- :heavy_check_mark: The change impacts existing ClassicPress users. (As stated in the original report, though I still don't know which specific plugin is being referenced here.)
- :question: The change is not going to break any other use cases. (This change would push us into a weird "in between" space where we provide more functions than WP 4.9 but less than WP 5.x.  I am not at all sure that this is a good idea.  For example, a plugin that tries to use this function will quite possibly just fail in some other way because they haven't checked for compatibility based on the WP version.)
- :white_check_mark: The change has automated tests. (From reading through the relevant WP code, it does, but just barely... they are just making sure the function does _something_ without causing errors, not really that it actually works as intended.)
- :x: I understand the change very well, or I can ask questions of someone who understands it very well. (Why did WP decide to add this function? What do they use it for? What other changes is this linked with?)

Overall I am not very comfortable with this change, and I think we should maintain the expectation of compatibility with WordPress 4.9 here.  In this case, plugins should be detecting whether the `rest_preload_api_request` function exists, or conditionally using it based on the `$wp_version` variable.  Plugins that don't do this would have the same issue on WP 4.9, and we have already set the expectation that we require WP 4.9 compatibility.

Also, perhaps most importantly, we'd be breaking what is perhaps the golden rule of programming: don't make a change unless you understand its effects.
